### ISSUE TITLE
[1.16] Mobs with a width of 0.9 (Cows, Pigs, Sheep, and Llamas in vanilla) don't move until they're moved manually by something else when spawned

### ISSUE THREAD

**Issue by seymourimadeit** [MinecraftForge/MinecraftForge#7730]
As the title says, adult pigs, sheep, llamas, and cows spawned won't move at all until they're forced to move.

Steps to reproduce:
1. Spawn an adult sheep, llama, pig or cow via a spawn egg
2. Watch the entity not move at all until something else forces it to move

Notes : This issue seems to only happen in version 1.16.4-35.1.9+ after 7520 was merged. The entities affected by this issue have a width of 0.9 defined in the EntityTypes class.

Video : 
Forge (no mods, 36.1.12):
https://imgur.com/a/oV9zU7g
Vanilla:
https://imgur.com/a/DGmCzoo

**Comment by Tslat** [comment]
I think I'm having this with mobs even when not spawned by a spawn egg

It does seem to relate to entity size though, not sure how yet

It seems like any amount of movement can cause it too, even tiny amounts

**Comment by OneLemonyBoi** [comment]
seems that this line:
`boolean flag = d0 < (double)this.maxDistanceToWaypoint && d2 < (double)this.maxDistanceToWaypoint && d1 < 1.0D;`
doesnt account for floating point error when casting to doubles, it should be:
`boolean flag = d0 <= (double)this.maxDistanceToWaypoint && d2 <= (double)this.maxDistanceToWaypoint && d1 < 1.0D;`
currently writing a PR to fix this.

**Comment by Lupicus** [comment]
Why not apply part of #6455 which was a fix to #6091 (i.e. #7520) instead?

**Comment by Tslat** [comment]
This still really needs to be looked at
It affects a lot of mobs

**Comment by LexManos** [comment]
The problem is every time I look into it, With just Forge. I can't reproduce it at all.
Mobs wander around just fine.
And just slapping things around with regard to the navigator is just asking to break more things.

**Comment by OneLemonyBoi** [comment]
im able to reproduce this on vanilla forge

**Comment by Tslat** [comment]
Keeping in mind the result is based on the mob's hitbox and starting position.
You won't see it on some mobs, you'll see it on others.

Even if it doesn't happen in vanilla, the fact that forge causes it to happen with modded mods should be enough to fix it

**Comment by LexManos** [comment]
Then make a proper writeup, with anything possible to make it reproduceable.
Exact worlds, exact entities, exact positions, exact versions of Minecraft and forge, Exact computer specs.
Again, this is not reproduceable by us, or as it seems by most people in the community.
Honestly, if it IS such an issue, it shows that the navigation system is fragile and needs to be fixed by Mojang, as the original 'issue' {mobs spinning} is a vanilla one.

**Comment by OneLemonyBoi** [comment]
INFO:

**Forge Version: 1.16 Commit 98bae44**
**RAM Allocated: 3GB (3641 MB In-Game)**
**World Seed: MinecraftForge**
**World Generation Type: Default**
**World Settings: Creative Peaceful**
**Entity: minecraft:sheep**

**Java Version: Amazon Corretto Version 16.0.2**

**Computer Specs:**

MacBook Pro (15-inch, 2018)
2.6 GHz 6-Core Intel Core i7
16 GB 2400 MHz DDR4
Radeon Pro 560X 4 GB

*Monitors*

![Screen Shot 2021-07-24 at 8 14 39 PM](https://user-images.githubusercontent.com/50998649/126886491-577f8571-3a06-4b36-94e8-00c91f330b2e.png)

**Sheep Info:**

**{SheepEntity@18351} "SheepEntity['Sheep'/209, l='ServerLevel[New World]', x=39.50, y=64.00, z=-55.50]"**

>  eatAnimationTick = 0
>  eatBlockGoal = {EatGrassGoal@18369} "EatGrassGoal"
>  inLove = 0
>  loveCause = null
>  age = 0
>  forcedAge = 0
>  forcedAgeTimer = 0
>  ambientSoundTime = -10
>  xpReward = 0
>  lookControl = {LookController@18370} 
>  moveControl = {MovementController@18371} 
>  jumpControl = {JumpController@18372} 
>  bodyRotationControl = {BodyController@18373} 
>  navigation = {GroundPathNavigator@18350} 
>  goalSelector = {GoalSelector@18374} 
>  targetSelector = {GoalSelector@18375} 
>  target = null
>  sensing = {EntitySenses@18376} 
>  handItems = {NonNullList@18377}  size = 2
>  handDropChances = {float[2]@18378} [0.085, 0.085]
>  armorItems = {NonNullList@18379}  size = 4
>  armorDropChances = {float[4]@18380} [0.085, 0.085, 0.085, 0.085]
>  canPickUpLoot = false
>  persistenceRequired = false
>  pathfindingMalus = {EnumMap@18381}  size = 3
>  lootTable = null
>  lootTableSeed = 0
>  leashHolder = null
>  delayedLeashHolderId = 0
>  leashInfoTag = null
>  restrictCenter = {BlockPos@18382} "BlockPos{x=0, y=0, z=0}"
>  restrictRadius = -1.0
>  attributes = {AttributeModifierManager@18383} 
>  combatTracker = {CombatTracker@18384} 
>  activeEffects = {HashMap@18385}  size = 0
>  lastHandItemStacks = {NonNullList@18386}  size = 2
>  lastArmorItemStacks = {NonNullList@18387}  size = 4
>  swinging = false
>  swingingArm = null
>  swingTime = 0
>  removeArrowTime = 0
>  removeStingerTime = 0
>  hurtTime = 0
>  hurtDuration = 0
>  hurtDir = 0.0
>  deathTime = 0
>  oAttackAnim = 0.0
>  attackAnim = 0.0
>  attackStrengthTicker = 0
>  animationSpeedOld = 0.0
>  animationSpeed = 0.0
>  animationPosition = 0.0
>  invulnerableDuration = 20
>  timeOffs = 4987.232
>  rotA = 0.015072665
>  yBodyRot = -15.091949
>  yBodyRotO = -15.091949
>  yHeadRot = -15.091949
>  yHeadRotO = -15.091949
>  flyingSpeed = 0.02
>  lastHurtByPlayer = null
>  lastHurtByPlayerTime = 0
>  dead = false
>  noActionTime = 1
>  oRun = 0.0
>  run = 0.0
>  animStep = 0.0
>  animStepO = 0.0
>  rotOffs = 0.0
>  deathScore = 0
>  lastHurt = 0.0
>  jumping = false
>  xxa = 0.0
>  yya = 0.0
>  zza = 0.0
>  lerpSteps = 0
>  lerpX = 0.0
>  lerpY = 0.0
>  lerpZ = 0.0
>  lerpYRot = 0.0
>  lerpXRot = 0.0
>  lyHeadRot = 0.0
>  lerpHeadSteps = 0
>  effectsDirty = false
>  lastHurtByMob = null
>  lastHurtByMobTimestamp = 0
>  lastHurtMob = null
>  lastHurtMobTimestamp = 0
>  speed = 0.0
>  noJumpDelay = 0
>  absorptionAmount = 0.0
>  useItem = {ItemStack@18388} "1 air"
>  useItemRemaining = 0
>  fallFlyTicks = 0
>  lastPos = {BlockPos@18389} "BlockPos{x=39, y=64, z=-56}"
>  lastClimbablePos = {Optional@18390} "Optional.empty"
>  lastDamageSource = null
>  lastDamageStamp = 0
>  autoSpinAttackTicks = 0
>  swimAmount = 0.0
>  swimAmountO = 0.0
>  brain = {Brain@18391} 
>  handlers = {LazyOptional[3]@18392} 
>  type = {EntityType@18393} "entity.minecraft.sheep"
>  id = 209
>  blocksBuilding = true
>  passengers = {ArrayList@18394}  size = 0
>  boardingCooldown = 0
>  vehicle = null
>  forcedLoading = false
>  level = {ServerWorld@16871} "ServerLevel[New World]"
>  xo = 39.5
>  yo = 64.0
>  zo = -55.5
>  position = {Vector3d@18395} "(39.5, 64.0, -55.5)"
>  blockPosition = {BlockPos@18389} "BlockPos{x=39, y=64, z=-56}"
>  deltaMovement = {Vector3d@20960} "(0.0, -0.0784000015258789, 0.0)"
>  yRot = -106.12393
>  xRot = 0.0
>  yRotO = -106.12393
>  xRotO = 0.0
>  bb = {AxisAlignedBB@18397} "AABB[39.05000001192093, 64.0, -55.94999998807907] -> [39.94999998807907, 65.29999995231628, -55.05000001192093]"
>  onGround = true
>  horizontalCollision = false
>  verticalCollision = true
>  hurtMarked = false
>  stuckSpeedMultiplier = {Vector3d@18398} "(0.0, 0.0, 0.0)"
>  removed = false
>  walkDistO = 0.0
>  walkDist = 0.0
>  moveDist = 0.0
>  fallDistance = 0.0
>  nextStep = 1.0
>  nextFlap = 1.0
>  xOld = 39.5
>  yOld = 64.0
>  zOld = -55.5
>  maxUpStep = 0.6
>  noPhysics = false
>  pushthrough = 0.0
>  random = {Random@18399} 
>  tickCount = 443
>  remainingFireTicks = -1
>  wasTouchingWater = false
>  fluidHeight = {Object2DoubleArrayMap@18400}  size = 2
>  wasEyeInWater = false
>  fluidOnEyes = null
>  invulnerableTime = 0
>  firstTick = false
>  entityData = {EntityDataManager@18401} 
>  inChunk = true
>  xChunk = 2
>  yChunk = 4
>  zChunk = -4
>  movedSinceLastChunkCheck = false
>  packetCoordinates = {Vector3d@18398} "(0.0, 0.0, 0.0)"
>  noCulling = false
>  hasImpulse = false
>  portalCooldown = 0
>  isInsidePortal = false
>  portalTime = 0
>  portalEntrancePos = null
>  invulnerable = false
>  uuid = {UUID@18402} "eb4bd6c9-28b2-419c-a776-64ec01b3141a"
>  stringUUID = "eb4bd6c9-28b2-419c-a776-64ec01b3141a"
>  glowing = false
>  tags = {HashSet@18404}  size = 0
>  forceChunkAddition = false
>  pistonDeltas = {double[3]@18405} [0.0, 0.0, 0.0]
>  pistonDeltasGameTime = 0
>  dimensions = {EntitySize@18406} "EntityDimensions w=0.9, h=1.3, fixed=false"
>  eyeHeight = 1.2349999
>  canUpdate = true
>  captureDrops = null
>  persistentData = null
>  isAddedToWorld = true
>  baseClass = {Class@5168} "class net.minecraft.entity.Entity"
>  capabilities = null
>  valid = true

Sheep Data according to /data get entity @e[limit=1,sort=nearest,type=sheep]:

Sheep has the following entity data: {Brain: {memories: {}}, HurtByTimestamp: 0, Attributes: [{Base: 0.08d, Name: "forge:entity_gravity"}, {Base: 16.0d, Modifiers: [{Amount: -0.04055830611068845d, Operation: 1, UUID: [I; 167792900, 380193506, -1125331168, -1588143754], Name: "Random spawn bonus"}], Name: "minecraft:generic.follow_range"}, {Base: 0.23000000417232513d, Name: "minecraft:generic.movement_speed"}], Invulnerable: 0b, FallFlying: 0b, ForcedAge: 0, PortalCooldown: 0, AbsorptionAmount: 0.0f, FallDistance: 0.0f, InLove: 0, CanUpdate: 1b, DeathTime: 0s, HandDropChances: [0.085f, 0.085f], PersistenceRequired: 0b, UUID: [I; -347351351, 682770844, -1485413140, 28513306], Age: 0, Motion: [0.0d, -0.0784000015258789d, 0.0d], Health: 8.0f, Color: 0b, LeftHanded: 0b, Air: 300s, OnGround: 1b, Rotation: [-106.12393f, -6.984519f], HandItems: [{}, {}], ArmorDropChances: [0.085f, 0.085f, 0.085f, 0.085f], Pos: [39.5d, 64.0d, -55.5d], Fire: -1s, ArmorItems: [{}, {}, {}, {}], CanPickUpLoot: 0b, Sheared: 0b, HurtTime: 0s} 

Observed Behavior: No Movement at All

Path ""Server thread"@15,736 in group "SERVER": RUNNING" took to get here:

> followThePath:211, PathNavigator
> tick:189, PathNavigator
> serverAiStep:685, MobEntity
> aiStep:2411, LivingEntity
> aiStep:488, MobEntity
> aiStep:115, AgeableEntity
> aiStep:51, AnimalEntity
> aiStep:132, SheepEntity
> tick:2158, LivingEntity
> tick:300, MobEntity
> tickNonPassenger:611, ServerWorld
> guardEntityTick:554, World
> tick:404, ServerWorld
> tickChildren:851, MinecraftServer
> tickServer:787, MinecraftServer
> tickServer:78, IntegratedServer
> runServer:642, MinecraftServer
> lambda$spin$0:232, MinecraftServer
> run:748, Thread

Console Log:
https://gist.github.com/OneLemonyBoi/343e9a846db1d5cf9dd8ac7e4849ca16

Please LMK if you need anything more!

**Comment by LexManos** [comment]
So, good news! Someone on our team was able to reproduce and debug it!
Turns out that there are other AI tasks that WILL make entities wander even if not pushed, and for some reason they were activating in my dev every time I tried to reproduce. So meh, entities apparently still wandered around they just took extra time in some cases.
Either way the fix seems to work, as it is a type precision issue. 
And it's been fixed in the linked commits.
